app/controllers/uploads_controller.rb

```diff
@@ -20,7 +20,8 @@ def create
       failed = load_csv.failed_instances
       @upload.check_for_headers
 
-      validation_warnings = failed.map(&:display_errors_with_row).sort
+      validation_warnings = failed.sort { |a, b| a.errors[:row]&.first.to_i <=> b.errors[:row]&.first.to_i }
```

Comment:
The safe navigation operator is unnecessary.  You will always be calling `.first` on an array (i.e. no chance it can be `nil`)

Reply:
removed